### PR TITLE
create a config facade

### LINKED ISSUE #15460: Create a small façade to break servo_config dependency from style
`servo_config` is a crate in components/config that pulls in a lot of random stuff.

In gecko mode, most of this is unnecessary. However, we pull this in for Stylo because the style crate needs to check if certain options are set.

Stylo can't set these options anyway. We should create a façade module in style that exposes servo_config things in Servo mode, and exposes dummy functions that just return a constant boolean in Gecko mode.

It would look something like

```rust
// in lib.rs
#[cfg(feature="servo")]
extern crate servo_config;

mod servo_config_facade;


// in servo_config_facade.rs
#[cfg(feature = "servo")]
fn viewport_enabled() -> bool {
 use servo_config::prefs::PREFS;
 PREFS.get("layout.viewport.enabled").as_boolean().unwrap_or(false)
}

#[cfg(feature = "gecko")]
fn viewport_enabled() -> bool {true}

// similar functions for each use of servo_config in the style crate


// in stylesheets.rs
// https://github.com/servo/servo/blob/6d1d4b57838d930e6ce39be720ef6c6dbb74af26/components/style/stylesheets.rs#L972
if servo_config_facade::viewport_enabled() {
 // do viewport thingies
}
```

When done, `./mach build -p style` and `./mach build-geckolib` should both work

### PR DESCRIPTION
**Pull request by karan1276**
Ref: #15460

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15460(github issue number if applicable).

- [x] These changes do not require tests because no new functionality is added. Existing test should cover all the changes.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15919)

### CODE REVIEW

**Comment by karan1276** [comment]
`./mach build-geckolib --verbose` gives an error i am not able to debug, it says:
`error: could not execute process `rustc -vV` (never executed)`

File: components/style/servo_config_facade.rs
Comment (canaltinova) [inline components/style/servo_config_facade.rs]: You've changed the `extern crate servo_config;` to `#[cfg(feature = "servo")] extern crate servo_config;` in components/style/lib.rs. 
In gecko build it tries to find this crate but fails. You can either add a cfg to this use statement or you can simply add related use statements inside servo functions.(as Manishearth wrote in issue example.)

File: components/style/servo_config_facade.rs
Comment (canaltinova) [inline components/style/servo_config_facade.rs]: I think adding a brief explanation about this module with `//!` comments would be better instead.

**Review by canova — COMMENTED** [review]
It looks like the problem is the use statement as I mentioned in the comment below.
Also nox suggested to make this name simply `config` instead of  `servo_config_facade`. Could you change this?

**Comment by KiChjang** [comment]
I'm still not comfortable calling it a facade.

**Comment by karan1276** [comment]
@KiChjang I am fine calling it anything. Maybe you and @Manishearth can talk and come to a conclusion ?

File: components/style/servo_config_facade.rs
Comment (karan1276) [inline components/style/servo_config_facade.rs]: aah, that explains it. I think ill add an cfg. Just curious, how did you debug that because error did not log anything

**Comment by KiChjang** [comment]
In the issue, @nox suggested to call it `config` instead, and there seemed to be no objections from anyone.

**Comment by karan1276** [comment]
Okay, cool :+1:

File: components/style/config.rs
Comment (canaltinova) [inline components/style/config.rs]: Rustdoc uses `//!` for module or crate level documentation. Could you convert `/* .. */` to that  please? With this change you can remove the `#[allow(missing_docs)]` attribute in the 
`#[allow(missing_docs)] pub mod config;` line. Also we will be able to see this comment in the doc.servo.org after this.

nit: also it would be wonderful if you can shorten these lines to match license length while you're here.

File: components/style/servo_config_facade.rs
Comment (canaltinova) [inline components/style/servo_config_facade.rs]: rustc should give you an error like this when you run `./mach build-geckolib`:
```
error[E0432]: unresolved import `servo_config`
  --> /Users/canova/projects/servo/components/style/config.rs:11:5
   |
11 | use servo_config;
   |     ^^^^^^^^^^^^ no `servo_config` in the root

error: aborting due to previous error
```
What OS do you use? If this didn't show up, maybe it was a rustc bug?

File: components/style/config.rs
Comment (canaltinova) [inline components/style/config.rs]: nit: I'm not comfortable with  these `#[allow(missing_docs)]` attributes but it's OK, I guess. At least we can move them to a separate line maybe?

**Comment by cbrewster** [comment]
This looks good, maybe address @canaltinova's comment about either documenting the settings or moving the `no_doc` to a new line
There some tidy failures:
```
Checking files for tidiness...
./components/style/matching.rs:27: use statement is not in alphabetical order
	expected: selectors::matching::AFFECTED_BY_PSEUDO_ELEMENTS
	found: config
./components/style/stylesheets.rs:22: use statement is not in alphabetical order
	expected: selectors::parser::SelectorList
	found: config
./components/style/traversal.rs:16: use statement is not in alphabetical order
	expected: selector_parser::RestyleDamage
	found: config
```

**Comment by KiChjang** [comment]
Please also remove the merge commit.

**Comment by karan1276** [comment]
Actually i tried to rebase 2-3 times but always mess it up. There are conflicts at same line for multiple commits. It gets really difficult to remove just the right change.

merge on the other hand just compares final results and points out all conflicts in one go. If it's not really a big issue i would suggest we go for a merge here.

**Comment by Wafflespeanut** [comment]
No, we don't allow merge commits in general because it's a result of an improper (or absence of) rebase. If you want help with it, please ask.

**Comment by karan1276** [comment]
Yes it will be awesome if you can help.

**Comment by canaltinova** [comment]
https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing

I think this guide is very good for explaining how to do rebase and squash. But for this specific situation here's how I would do it:

First you need to delete this merge commit and squash others with interactive rebase:
`git rebase -ip HEAD~9`
`-i` tells that you want to use interactive rebase, and `-p` tells to preserve merge commits.
This will open an editor that you can manipulate the commits. You can delete the merge commit with deleting its line. And you should write the `fixup` or `f` keyword instead of `pick` to other commits except the first commit. Fixup will merge the commits into previous ones and they will become one.
When you save and exit, rebase will take care of it. And you will have just one commit. After that, you will need to fetch and rebase again like this:
`git pull upstream master --rebase` (you should replace upstream with your servo/servo remote name)
During rebase, some conflicts will happen. You should [resolve the conflicts](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing#fixing-conflicts) and `git add <file>` them. After that, you can write `git rebase --continue` and you are done.

Let us know if you still have questions. Or we can do it for you if you check the "Allow edits from maintainers" checkbox.

**Comment by karan1276** [comment]
@canaltinova Thanks! That's a good strategy. I won't follow it exactly but i will first squash and then rebase. I don't want to do them together because i suspect squash will have it's own conflicts(maybe not, not sure) and then rebase will defiantly have conflicts. And it's easier for me to handle them separately.

Thanks again

**Comment by canaltinova** [comment]
Yes, you can probably try to delete merge commit with `git rebase -ip HEAD~2` first and then proceed to squash them. I don't think you will have merge conflicts during squash but it would be easier.

**Comment by jdm** [comment]
`./mach test-tidy` says:
```
./components/style/matching.rs:26: use statement is not in alphabetical order
	expected: selectors::matching::AFFECTED_BY_PSEUDO_ELEMENTS
	found: config
./components/style/stylesheets.rs:31: use statement is not in alphabetical order
	expected: selectors::parser::SelectorList
	found: config
./components/style/traversal.rs:16: use statement is not in alphabetical order
	expected: selector_parser::RestyleDamage
	found: config
```

**Comment by karan1276** [comment]
These conflicts will pop up every time there is a merge in master. I would resolve them at the end just before this PR gets merged. Are there any other outstanding issues except for the conflicts?

**Comment by cbrewster** [comment]
this looks good if you address @canaltinova's comment about converting `/* ... */` to `//!` and rebase/squash

**Comment by karan1276** [comment]
some changes in #16297 are pretty major, ill remove conflicts but be cautious while reviewing

File: components/style/context.rs
Comment (canaltinova) [inline components/style/context.rs]: It looks like you removed the function header here.

File: components/style/context.rs
Comment (canaltinova) [inline components/style/context.rs]: This looks wrong. Currently servo returns `opts::get().style_sharing_stats` and gecko returns `*DUMP_STYLE_STATISTICS`. We need to preserve that behavior. You should remove the `*DUMP_STYLE_STATISTICS ||` from here and return `*DUMP_STYLE_STATISTICS` in gecko's style_sharing_stats_enabled function.

File: components/style/matching.rs
Comment (canaltinova) [inline components/style/matching.rs]: There is a merge conflict here.

File: components/style/traversal.rs
Comment (canaltinova) [inline components/style/traversal.rs]: Why did you remove `traversal_flags.for_reconstruct()`?

**Review by canova — CHANGES_REQUESTED** [review]
Oh, it looks like https://github.com/servo/servo/pull/16297/commits/c857878ce9479cf0c719d9bc3f922c6cb9cdcd8b implemented something like this. Sorry about that :/ I don't know whether we should still continue to this or not.

**Comment by karan1276** [comment]
> Oh, it looks like c857878 implemented something like this. Sorry about that :/ I don't know whether we should still continue to this or not.

Yup, you are probably right. Maybe you can tell me for sure if this pr is needed anymore ? If not then lets just shut it right away, why spend more time.

**Comment by cbrewster** [comment]
Closing as it looks like this is already implemented.

**Pull request closed without merging**
